src/scheduler/mod.rs

```diff
@@ -78,14 +82,159 @@ pub fn sched_yield(current_stack: &'static Context) -> &'static Context {
+            match elem.quanta {
+                x if x <= current_thread.quanta => {
+                    current_thread.quanta -= elem.quanta;
+                    true
+                },
+                _ => false,
+            }
+        }).count();
+    // first, update the delta for the element following, if it exists
+    if let Some(next) = sleeping.get_mut(index) {
+        next.quanta -= current_thread.quanta;
+    }
+    // now lets put it in the queue
+    sleeping.insert(index, current_thread);
+
+    ret
+}
+
+/// Reduce the current thread's time slice by one tick. If it has no
+/// time left then yield to a new thread.
+pub fn tick(current_stack: &'static Context) -> &'static Context {
+    let mut lock = SCHEDULER.lock();
+    let &mut Scheduler {
+        ref mut threads,
+        ref mut sleeping,
+        ref mut current,
+        ref mut idle,
+    } = lock.as_mut().unwrap();
+
+    // update the sleeping thread list
+    if let Some(thread) = sleeping.front_mut() {
+        thread.quanta -=1;
+    }
+    loop {
+        let should_pop = sleeping.front()
+            .map_or(false, |thread| thread.quanta == 0);
+        if should_pop {
+            threads.push_back(sleeping.pop_front().unwrap());
+        } else {
+            break;
+        }
+    }
+
+    // now update the running thread
+    {
+        let mut running = current.as_mut().unwrap_or(idle);
+
+        running.quanta -= 1;
+        if running.quanta > 0 {
+            // continue with the current thread
+            return current_stack;
+        }
+    }
+    let mut next_thread = threads.pop_front();
+
+    if next_thread.is_none() && current.is_none() {
+        // Only the idle thread can run
+        idle.quanta = TICKS;
+        return current_stack;
+    }
+
+    // Now swap threads
+    let ret = {
+        if let Some(current_thread) = current {
+            // `next_thread` is unknown and current is running
+            let next = next_thread.as_mut().unwrap_or(idle);
+            current_thread.swap(current_stack, next)
+        } else {
+            // `next_thread` must be Some(_) and idle is running
+            idle.swap(current_stack, next_thread.as_mut().unwrap())
+        }
+    };
 
+    if let Some(current) = current.take() {
+        threads.push_back(current);
+    }
+
+    *current = next_thread;
     ret
 }
 
 /// Reschedule the current kernel thread
 pub fn thread_yield() {
     unsafe {
-        asm!("int $0" :: "i"(YIELD_INT) :: "volatile")
+        asm!("mov rax, 0
+              int $0"
+              :: "i"(SLEEP_INT)
+              : "rax"
+              : "intel", "volatile")
+    }
+}
+
+pub fn thread_sleep(time: u8) {
+    unsafe {
+        asm!("movzx rax, $1
+              int $0"
+              :: "i"(SLEEP_INT),"r"(time)
+              : "rax"
+              : "intel", "volatile")
+    }
+}
+
+pub fn test() {
+    add(A);
+    add(B);
+    thread_sleep(255);
+    println!("Re-entered the main thread!");
+}
```

Comment:
Ok, so let's have the discussion on testing this. There aren't really any assertions in this code, so it needs to be slightly modified. 

Moreover, I guess the bigger issue is what if some assertion isn't reached, but the build passes because tests are counted lazily. If we think this is a problem, we can switch our philosophy to having a magic number, or we could try to break this into multiple test groups - although tappy may not like that. Or we could try to rig up some other sort of failsafe.

Reply:
There isn't some way we could break up the input from serial somehow? I know this might not be the prettiest, but couldn't we write tap output to a file, and then we can either write it to multiple files or write it to one file and break it up into multiple files, and then consume the files with tappy?

Reply:
I mean, I could probably hack something together but it wouldn't be pretty

Reply:
Tell me what you think of afb14b6

Reply:
tests are added